sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
   }
 
   /**
+   * Statically evaluate an expression containing one or more aggregates on an empty input.
+   */
+  private def evalOnZeroTups(expr : Expression) : Option[Any] = {
+    // AggregateExpressions are Unevaluable, so we need to replace all aggregates
+    // in the expression with the value they would return for zero input tuples.
+    val rewrittenExpr = expr transform {
+      case a @ AggregateExpression(aggFunc, _, _, resultId) =>
+        val resultLit = aggFunc.defaultResult match {
+          case Some(lit) => lit
+          case None => Literal.default(NullType)
+        }
+        Alias(resultLit, "aggVal") (exprId = resultId)
+    }
+    Option(rewrittenExpr.eval())
+  }
+
+  /**
    * Construct a new child plan by left joining the given subqueries to a base plan.
    */
   private def constructLeftJoins(
       child: LogicalPlan,
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
+        val aggOutputExpr = query.asInstanceOf[Aggregate].aggregateExpressions.head
```

Comment:
Upon further reflection, a Filter node above the Aggregate creates additional issues. If you rewrite the subquery into an outer join above the Filter, it may be impossible to distinguish between two cases:
- The original subquery returns null because an aggregate value did not pass the HAVING clause
- The aggregate in the original subquery is evaluated over zero tuples

In both cases, the outer join will return a tuple containing nulls.
For example, the query:
```sql
select * from l where (select count(*) cnt from r where l.a = r.c having cnt = 0) = 0
```
would turn into
```sql
select * 
from l left outer join (select c, count(*) cnt from r group by c having cnt = 0) sq
    on l.a = sq.c 
where sq.cnt = 0
```
Note how the rewritten subquery `select c, count(*) cnt from r group by c having cnt = 0` always returns zero tuples.

I'm working through a few possible solutions, trying to see if one of them is guaranteed to be correct.